obsproject/obs-studio
# libobs, obs-x264: MISP precision timestamps

### Description



Introduce MISP precision timestamps, based on the following open standards:

- https://www.itu.int/rec/T-REC-H.264
- https://gwg.nga.mil/misb/docs/standards/ST0603.5.pdf
- https://gwg.nga.mil/misb/docs/standards/ST0604.6.pdf

This PR attaches MISP precision timestamps to the x264 encoder to introduce this new module. There is a follow-up PR ready which will attach MISP timestamps to Lain's NVENC encoder. Based on this design, attaching MISP timestamps to _any_ OBS encoder should only require a few lines of code.

### Motivation and Context




This change allows downstream players to synchronise incoming OBS streams for which there are countless applications. One of those would be esports tournaments where multiple people observe matches and stream their observations into a production control room. The ability to synchronise these incoming streams means that switching between streams would not create unwanted "jumps in time".

### How Has This Been Tested?




The implementation of the specification has been tested on Mac and Windows by streaming to an RTMP endpoint and reading from it using `ffmpeg`:

```bash
ffmpeg -i [...] -c:v copy -bsf:v trace_headers -f null -
```

- Mac: BigSur 11.1 on a MacBook Pro 8GB
- Windows: Windows 10 on an AWS' `g4dn.2xlarge`

The exact tests that were run:
- Enable/disbale sending of timestamps before/during streaming
- Update UTC<>TAI difference before/during streaming
- Change system time before/during streaming (observe attaching the correct direction of discontinuty)

### Types of changes








- New feature (non-breaking change which adds functionality)

### Checklist:
